### PR TITLE
Fixes #34374 - Import/Export based on repository

### PR DESCRIPTION
**Pull request by parthaa**
### Setup
- Check out the PR. 
- `rake db:migrate`
- Start your foreman server 

-  Go to your hammer environment.
- Checkout https://github.com/Katello/hammer-cli-katello/pull/836
```
cd hammer-cli-katello
bundle install
bundle exec bash
```
### Export Seed
- Run the seeder script. This script creates export org, products, repositories, content-views and exports the created cv version
```bash
#!/bin/bash
shopt -s expand_aliases
set -e
set -x
EXPORT_ORG="export-$RANDOM"
PRODUCT="prod"
REPO1="katello"
REPO1_URL="https://fedorapeople.org/groups/katello/releases/yum/3.8/katello/el7/x86_64/"
REPO2="misc"
REPO2_URL="https://partha.fedorapeople.org/test-repos/pteradactly-with-dino-errata/"
REPO3="candlepin"
REPO3_URL="https://fedorapeople.org/groups/katello/releases/yum/3.8/candlepin/el7/x86_64/"
CONTENT_VIEW="view"

# create import export orgs
hammer organization create --name=$EXPORT_ORG

#create the products in both import and export orgs
hammer product create --name=$PRODUCT --organization=$EXPORT_ORG

#create and sync repositories in export org
hammer repository create --content-type=yum --product=$PRODUCT  --name=$REPO1 --url=$REPO1_URL --organization=$EXPORT_ORG
hammer repository create --content-type=yum --product=$PRODUCT  --name=$REPO2 --url=$REPO2_URL --organization=$EXPORT_ORG
hammer repository create --content-type=yum --product=$PRODUCT  --name=$REPO3 --url=$REPO3_URL --organization=$EXPORT_ORG

hammer repository synchronize --product=$PRODUCT  --name=$REPO1 --organization=$EXPORT_ORG
hammer repository synchronize --product=$PRODUCT  --name=$REPO2 --organization=$EXPORT_ORG
hammer repository synchronize --product=$PRODUCT  --name=$REPO3 --organization=$EXPORT_ORG

# create content view in EXPORT_ORG
hammer content-view create --name=$CONTENT_VIEW --organization=$EXPORT_ORG
hammer content-view add-repository --name=$CONTENT_VIEW --repository=$REPO1 --product=$PRODUCT --organization=$EXPORT_ORG
hammer content-view add-repository --name=$CONTENT_VIEW --repository=$REPO2 --product=$PRODUCT --organization=$EXPORT_ORG
hammer content-view add-repository --name=$CONTENT_VIEW --repository=$REPO3 --product=$PRODUCT --organization=$EXPORT_ORG
hammer content-view publish --name=$CONTENT_VIEW  --organization=$EXPORT_ORG
hammer content-view version list --content-view=view  --organization=export-5534 
```
- The above script generates 1 content view version. v1 has 3 repos
- Library has 1 product
- You should get an output that looks like

```
$ hammer content-view version list --content-view=view  --organization=export-5534 
---|----------|---------|-------------|-----------------------
ID | NAME     | VERSION | DESCRIPTION | LIFECYCLE ENVIRONMENTS
---|----------|---------|-------------|-----------------------
3  | view 1.0 | 1.0     |             | Library               
---|----------|---------|-------------|-----------------------

```

 

```
$ hammer content-export repository --name=katello --organization=export-5534  --product=prod
[.........................................................................................................................................................................................................] [100%]
Generated /var/lib/pulp/exports/export-5534/view/1.0/2020-11-18T20-05-41-00-00/metadata.json
```

- Do an ls on the export directory and you should see a `tar.gz`, a `<id>toc.json` and a metadata.json
```
$ ls -lh /var/lib/pulp/exports/export-5534/view/1.0/2020-11-18T20-05-41-00-00/
total 68M
-rw-rw-r--+ 1 pulp    pulp    68M Nov 18 20:05 export-3d3a0387-d0b6-41fe-8c95-a8d60adf7722-20201118_2005.tar.gz
-rw-rw-r--+ 1 pulp    pulp    333 Nov 18 20:05 export-3d3a0387-d0b6-41fe-8c95-a8d60adf7722-20201118_2005-toc.json
-rw-rw-r--+ 1 vagrant vagrant 432 Nov 18 20:05 metadata.json
```

### What Does the Export operation do
- It tries to fetch a content view with the name `Export-<Repository Label>` or `Export-<Repository Label>-#{destination_server}` in the org specified if you provided a destination_server name. If it can't find such a content view it creates a new one with that name.
- Marks the content view as `generated_by_export`
- Adds this repository to `Export-<Repository Label>` content view
- Publishes it
- Generates an export out of the published version.

### What Else
- Adds a repository subcommnd  to `incremental` end point. Use this if you want to increment from the previous version of the repository. 
- See the example below
```
$ hammer content-export incremental repository --name=katello --organization=export-5534  --product=prod
[.....................................................................................................................................................................................................................................] [100%]
Generated /var/lib/pulp/exports/export-5534/Export-Library/2.0/2020-11-28T22-32-47-00-00/metadata.json
$ ls -lh  /var/lib/pulp/exports/export-5534/Export-Library/2.0/2020-11-28T22-32-47-00-00/ 
total 176K
-rw-rw-r--+ 1 pulp    pulp    166K Nov 28 22:32 export-d31e5bbe-c321-46cd-ac53-15c470ecfe1b-20201128_2232.tar.gz
-rw-rw-r--+ 1 pulp    pulp     333 Nov 28 22:32 export-d31e5bbe-c321-46cd-ac53-15c470ecfe1b-20201128_2232-toc.json
-rw-rw-r--+ 1 vagrant vagrant  496 Nov 28 22:32 metadata.json
```
-  The size of the export tgz should be much smaller because we asked only for an incremental export and nothing changed in the repository. 

### What Does the Import operation do
- It tries to fetch a content view with the name `Import-<Repository Label>` in the import org. If it can't find such a content view it creates a new view with that name.
- Adds the appropriate repo based on the metadata (also marked the cv import-only and generated-by-export)
- Runs a version import into this new content view and updates the library accordingly. 

- Create the import org
```
hammer organization create --name=my-import-org
```
 
- Run the import command below (change org and path to appropriate value)

```
$ hammer content-import repository --organization=my-import-org --path=/var/lib/pulp/exports/export-20214/Export-Library/3.0/2020-12-02T15-39-54-00-00/
[.....................................................................................................................................................................................................................................] [100%]
```

- You should notice a new repository created in the import org and content in that repository must be updated
- You should notice a new Content View  `Import-<Repository Label>`.

### Things to test

- Play with both complete and incremental exports of repository.
- Play with imports
- Go to content views in the UI
- Content views generated by repository export/import will not be visible to the user. However export-library and import-library will be. This is to address customers who find value in entire library snapshots and want  to make use of it.
- Disabling/Removing a repository generated content view should result in the content view getting destroyed.
- You should not be able to get a host to subscribe- pull content  from repository generated content view.
- Capsule sync will not sync any repository generated content view.

cv-list-page
- You should not be able to publish/add repositories or add filters for the generated repositories
Content-View-Details-Repos
Content-View-Details

### CODE REVIEW

File: app/controllers/katello/api/v2/content_export_incrementals_controller.rb [hunk 1]
```diff
@@ -87,5 +115,12 @@ def find_exportable_organization
         throw_resource_not_found(name: 'organization', id: params[:organization_id])
       end
     end
+
+    def find_exportable_repository
+      @repository = Repository.find_by_id(params[:id])
+      if @repository.blank? || !@repository.organization.can_export_content?
```
Comment (jturel) [inline app/controllers/katello/api/v2/content_export_incrementals_controller.rb:121]: May want to throw a separate error for can_export_content? evaluating to false

File: app/lib/actions/katello/organization/destroy.rb [hunk 2]
```diff
@@ -61,7 +61,7 @@ def remove_environments(organization)
               plan_action(Katello::Environment::Destroy, env, :skip_repo_destroy => true, :organization_destroy => true)
             end
           end
-          plan_action(Katello::Environment::Destroy, organization.library, :skip_repo_destroy => true, :organization_destroy => true)
+          plan_action(Katello::Environment::Destroy, organization.library, :skip_repo_destroy => true)
```
Comment (jturel) [inline app/lib/actions/katello/organization/destroy.rb:64]: What's this for?

File: app/models/katello/repository.rb [hunk 3]
```diff
@@ -302,7 +302,7 @@ def on_demand?
     end
 
     def immediate?
-      root.download_policy == ::Katello::RootRepository::DOWNLOAD_ON_IMMEDIATE
+      root.download_policy == ::Katello::RootRepository::DOWNLOAD_IMMEDIATE
```
Comment (jturel) [inline app/models/katello/repository.rb:305]: Why change this?

Comment (parthaa) [reply app/models/katello/repository.rb:305]: Oh that was a buggy code. Some one changed the constant name from  `DOWNLOAD_ON_IMMEDIATE`  to `DOWNLOAD_IMMEDIATE` but forgot to change it in the model here

Comment (jturel) [reply app/models/katello/repository.rb:310]: Actually I don't see any code calling "immediate?" so maybe it's not urgent

Comment (parthaa) [reply app/models/katello/repository.rb:310]: https://github.com/Katello/katello/pull/9925/files#diff-6dd39994c7a6695008f92af2b9727cabe5ce0f3aa74e4b00776bab6f7f228b2cR42
look at `validate_repositories_immediate` in `export_repository` action

Comment (jturel) [reply app/models/katello/repository.rb:310]: ACK. Last thing - is this change compatible with 6.10 ie when did the constant reference get changed/broken ?

File: app/services/katello/pulp3/content_view_version/metadata_generator.rb [hunk 4]
```diff
@@ -17,9 +17,10 @@ def initialize(export_service:)
         def generate!
           ret = { organization: organization.name,
                   repositories: {},
-                  content_view: content_view.slice(:name, :label, :description),
+                  content_view: content_view.slice(:name, :label, :description, :generated_for),
```
Comment (jturel) [inline app/services/katello/pulp3/content_view_version/metadata_generator.rb:20]: It's inevitable that we'll be changing the metadata from time to time and  imo we need to make sure that any exports that exist out there today will continue to import without issue - those would be missing generated_for and destination_server per this change

Comment (parthaa) [reply app/services/katello/pulp3/content_view_version/metadata_generator.rb:20]: Should be fixed!

File: app/controllers/katello/api/v2/organizations_controller.rb [hunk 5]
```diff
@@ -134,7 +134,7 @@ def releases
 
     api :PUT, "/organizations/:id/cdn_configuration", N_("Update the CDN configuration")
     param :id, String, :desc => N_("ID of the Organization"), :required => true
-    param :type, String, :desc => N_("CDN configuration type. One of %s.") % CdnConfiguration::TYPES, :required => true
+    param :type, String, :desc => N_("CDN configuration type. One of %s.") % CdnConfiguration::TYPES.join(", "), :required => true
```
Comment (jturel) [inline app/controllers/katello/api/v2/organizations_controller.rb:137]: This change isn't compatible with 6.10, so it'll raise a merge conflict downstream. Not a big deal i guess..

**Comment by jturel** [comment]
@parthaa this all looks great!

regarding the enum for generated_for I think all the values are OK even though we don't expressly depend on them. it may help in the future to have that information populated already.

only quirk I've seen is that registering with subman reports the repository environment which should be hidden:

```
Registering to: centos7-katello-devel.jturel.example.com:443/rhsm                                                                                                                             
Username: admin                                                                                                                                                                               
Password:                                                                                                                                                                                     
Hint: User "admin" is member of following organizations: Default_Organization, Test2, export-1, import                                                                                        
Organization: import                                                                                                                                                                          
Hint: Organization "import" contains following environments: Library, **Library/Import-katello-74**                                                                                               
Environment: Library                                                                                                                                                                          
The system has been registered with ID: 5e1c7238-8534-4c8a-9002-498d25ffe125                                                                                                                  
The registered system name is: centos7-katello-client.jturel.example.com   
```

**Comment by parthaa** [comment]
> only quirk I've seen is that registering with subman reports the repository environment which should be hidden:

Updated :) . check last commit

**Review by jturel — APPROVED** [review]
APJ!

**Pull request merged**